File: Low Level Design (LLD)/DesignPatterns/3. Structural/Visitor/Java/Directory.java
import java.util.ArrayList;
import java.util.List;

public class Directory implements Element {
    private String name;
    private List<Element> children = new ArrayList<>();

    public Directory(String name) {
        this.name = name;
    }

    public void add(Element element) {
        children.add(element);
    }

    public String getName() {
        return name;
    }

    public List<Element> getChildren() {
        return children;
    }

    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
        for (Element e : children) {
            e.accept(visitor);
        }
    }
}

File: Low Level Design (LLD)/DesignPatterns/3. Structural/Visitor/Cpp/src/Directory.cpp
#include "../include/Directory.h"
#include "../include/Visitor.h"

Directory::Directory(const std::string& name) : name(name) {}

Directory::~Directory() {
    for (auto child : children) {
        delete child;
    }
}

std::string Directory::getName() const {
    return name;
}

void Directory::add(Element* element) {
    children.push_back(element);
}

void Directory::accept(Visitor* visitor) {
    visitor->visit(this);
    for (auto child : children) {
        child->accept(visitor);
    }
}
